### PR TITLE
[Core] Adding token ranks along with logprobs

### PR DESCRIPTION
**Pull request by SwapnilDreams100**
Adds the token rank functionality (for both prompt tokens and sample tokens) within the `Logprobs` object, by adding the `rank` property to `Logprobs`. 
This PR also adds a couple of tests to verify the ranks in greedy and non-greedy settings.

The implementation here is meant to align with functionality in https://github.com/IBM/text-generation-inference (IBM's fork of [HF TGI](https://github.com/huggingface/text-generation-inference)).

### CODE REVIEW

File: vllm/model_executor/layers/sampler.py [hunk 1]
```diff
@@ -447,6 +447,9 @@ def _sample(
     ]
     return sample_results
 
+def _get_ranks(x: torch.Tensor, indices: List[int]) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (SwapnilDreams100) [reply vllm/model_executor/layers/sampler.py:451]: will do!

File: vllm/sequence.py [hunk 2]
```diff
@@ -17,9 +17,9 @@
 class Logprob:
     """Infos for supporting OpenAI compatible logprobs."""
     logprob: float
+    rank: Optional[int] = None
```
[the remark this answers is not in the document]
Comment (SwapnilDreams100) [reply vllm/sequence.py:20]: will do!

**Review by njhill — APPROVED** [review]
Thanks @SwapnilDreams100!

File: vllm/model_executor/layers/sampler.py [hunk 3]
```diff
@@ -490,6 +508,9 @@ def _get_logprobs(
         batched_logprobs_query_seq_indices,
         batched_logprobs_query_token_indices
     ]]
+    batched_ranks_query_result = _get_ranks(
+        logprobs[batched_logprobs_query_seq_indices],
+        batched_logprobs_query_token_indices)
```
Comment (Yard1) [inline vllm/model_executor/layers/sampler.py:513]: let's move this after line 526, since we'll have to force a CPU<->GPU sync here anyway.

**Review by Yard1 — APPROVED** [review]
This looks good, please move the invocation as outlined in the comment before merge!

**Comment by njhill** [comment]
One question I guess is whether these should also be exposed in the openai API responses (but don't necessarily need to be addressed in this PR).

File: vllm/model_executor/layers/sampler.py (hunk 3 above)
Comment (SwapnilDreams100) [reply vllm/model_executor/layers/sampler.py]: will do!

**Review by rkooo567 — APPROVED** [review]

**Pull request merged**